dsa/src/signing_key.rs

```diff
@@ -80,31 +87,57 @@ impl SigningKey {
     where
         D: Digest + BlockSizeUser + FixedOutputReset,
     {
-        let k_kinv = crate::generate::secret_number_rfc6979::<D>(self, prehash);
+        let k_kinv = crate::generate::secret_number_rfc6979::<D>(self, prehash)?;
         self.sign_prehashed(k_kinv, prehash)
     }
 
     /// Sign some pre-hashed data
     fn sign_prehashed(
         &self,
-        (k, inv_k): (BigUint, BigUint),
+        (k, inv_k): (BoxedUint, BoxedUint),
         hash: &[u8],
     ) -> signature::Result<Signature> {
+        extern crate std;
+
         let components = self.verifying_key().components();
+        let key_size = &components.key_size;
         let (p, q, g) = (components.p(), components.q(), components.g());
         let x = self.x();
 
-        let r = g.modpow(&k, p) % q;
+        debug_assert_eq!(key_size.n_aligned(), q.bits_precision());
+
+        let x = x.widen(p.bits_precision());
+        let x = &x;
+
+        let k = k.widen(p.bits_precision());
+        let inv_k = inv_k.widen(p.bits_precision());
+
+        let params = BoxedMontyParams::new(p.clone());
+        let form = BoxedMontyForm::new((**g).clone(), params);
+        let r = form.pow(&k).retrieve() % q.widen(p.bits_precision());
+        debug_assert_eq!(key_size.l_aligned(), r.bits_precision());
+
+        let r_short = r.shorten(key_size.n_aligned());
+        // TODO(baloo): is there any way R could be zero here? Could it be any reason for K to be
+        //              one?
+        let r_short = NonZero::new(r_short).unwrap();
+        let r = NonZero::new(r).unwrap();
```

Comment:
I'm not sure about this unwrap, isn't there a way for `r = g.modpow(k, p) % q` to be zero?

Reply:
Oh right, that ties to the `k` value computed in the `secret_number` and `secret_number_rfc6979`!

That would then be a bug and a panic might be in order.